### PR TITLE
[`ConstantLengthDataset`] Fix packed dataset issue

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes https://github.com/lvwerra/trl/issues/450 

As reported in that issue, there is an inconsistent behaviour when using `packing=True` and the number of iterations showed in the logs of training. In fact, the inconsistency comes from the fact that in the packed case, sequences are packed all together on the fly until `max_seq_length` is reached, leading to reducing the number of total expected samples.

The fix is to force-set the `infinite` argument to `True` if a user explicity decides to train a model with `max_steps` strategy but properly warn them that the argument has been overriden and warn the user when the dataset is moved to the next iteration.

In the case of epochs training strategy, we should force-set infinite argument to False, otherwise the training will run forever.

Added also nice CI tests to make sure that this behavior will stay consistent in future commits.  Added also a line in the documentation.

cc @lvwerra @vwxyzjn

### CODE REVIEW

File: docs/source/sft_trainer.mdx [hunk 1]
```diff
@@ -98,6 +98,8 @@ trainer = SFTTrainer(
 trainer.train()
 ```
 
+Note that if you use a packed dataset and if you pass `max_steps` in the training arguments you will probably train your models for more than few epochs.
```
Comment (lvwerra) [inline docs/source/sft_trainer.mdx:101]: That's not really true, right? It could also be shorter - depends on the dataset and max_steps you choose :)

File: trl/trainer/sft_trainer.py [hunk 2]
```diff
@@ -219,6 +219,17 @@ def __init__(
             preprocess_logits_for_metrics=preprocess_logits_for_metrics,
         )
 
+        if self.args.max_steps > 0 and packing:
+            warnings.warn(
+                "You passed `packing=True` to the SFTTrainer, and you are training your model with `max_steps` strategy. Will force-set `infinite` argument of the packed dataset to `True` "
```
Comment (lvwerra) [inline trl/trainer/sft_trainer.py:224]: Since most users might not know the infinite setting maybe we can say something like:
```suggestion
                "You passed `packing=True` to the SFTTrainer, and you are training your model with `max_steps` strategy. The dataset will be iterated until the `max_steps` are reached."
```

File: trl/trainer/sft_trainer.py [hunk 3]
```diff
@@ -219,6 +219,17 @@ def __init__(
...
+            )
+            self.train_dataset.infinite = True
+        elif self.args.max_steps == -1 and packing:
+            warnings.warn(
```
Comment (lvwerra) [inline trl/trainer/sft_trainer.py:228]: I think in that case we don't have to warn - the model will train as expected, no?

File: trl/trainer/utils.py [hunk 4]
```diff
@@ -244,6 +244,7 @@ def __iter__(self):
                 except StopIteration:
                     if self.infinite:
                         iterator = iter(self.dataset)
+                        warnings.warn("The dataset reached end. The iterator is reset")
```
Comment (lvwerra) [inline trl/trainer/utils.py:247]:
```suggestion
                        warnings.warn("The dataset reached end and the iterator is reset to the start.")
```

File: tests/test_sft_trainer.py [hunk 5]
```diff
@@ -423,6 +423,65 @@ def test_data_collator_completion_lm(self):
         result_text = self.tokenizer.decode(batch["input_ids"][0, last_pad_idx + 1 :])
         self.assertTrue(result_text == "I have not been masked correctly.")
 
+    def test_sft_trainer_infinite_with_model(self):
```
Comment (lvwerra) [inline tests/test_sft_trainer.py:426]: What do you think about double checking here for both cases if we trained as many steps as we expect?

File: trl/trainer/sft_trainer.py (hunk 3 above)
Comment (younesbelkada) [reply trl/trainer/sft_trainer.py]: Makes sense, removed it

**Review by lvwerra — APPROVED** [review]
